File: src/test/java/org/artembogomolova/demo/webapp/test/domain/db/dao/auth/AuthorityDaoTest.java
package org.artembogomolova.demo.webapp.test.domain.db.dao.auth;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.artembogomolova.demo.webapp.dao.repo.auth.IAuthorityRepository;
import org.artembogomolova.demo.webapp.domain.auth.Authority;
import org.artembogomolova.demo.webapp.domain.auth.Authority_;
import org.artembogomolova.demo.webapp.domain.auth.Role;
import org.artembogomolova.demo.webapp.test.domain.db.dao.AbstractDaoTest;
import org.artembogomolova.demo.webapp.validation.UniqueMultiColumn.UniqueMultiColumnConstraint;
import org.junit.jupiter.api.DisplayName;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.repository.CrudRepository;

@DisplayName("Dao test: Authority")
class AuthorityDaoTest extends AbstractDaoTest<Authority> {

  private static final String NAME_VALUE = "test:authority";
  private static final String DESCRIPTION_VALUE = "test description";
  private static final String NAME_ANOTHER_VALUE = "test:authority_another";
  @Autowired
  private IAuthorityRepository authorityRepository;

  AuthorityDaoTest() {
    super(Authority.class, Authority::new);
  }

  @Override
  protected List<Authority> updateEntities(List<Authority> savedCollection) {
    savedCollection.get(0).setDescription(DESCRIPTION_VALUE + "1");
    return savedCollection;
  }

  @Override
  protected CrudRepository<Authority, Long> getCrudRepository() {
    return authorityRepository;
  }

  @Override
  protected List<Authority> generateEntities() {
    Authority result = buildAuthority();
    return List.of(result);
  }

  private Authority buildAuthority() {
    Authority result = new Authority();
    result.setName(NAME_VALUE);
    result.setDescription(DESCRIPTION_VALUE);
    return result;
  }

  @Override
  protected Map<String, Object> buildCommonFieldValues(UniqueMultiColumnConstraint uniqueMultiColumnConstraint) {
    Map<String, Object> result = new HashMap<>();
    result.put(Authority_.ROLES, new ArrayList<>());
    return result;
  }

  @Override
  protected Authority doPrepareDeniedTestEntity(UniqueMultiColumnConstraint uniqueMultiColumnConstraint, Map<String, Object> commonValues) {
    return buildAuthority(commonValues);
  }

  @Override
  protected Authority doDuplicateDeniedTestEntity(UniqueMultiColumnConstraint columns, Map<String, Object> commonValues) {
    return buildAuthority(commonValues);
  }

  private Authority buildAuthority(Map<String, Object> commonValues) {
    Authority result = buildAuthority();
    //result.setRoles((List<Role>) commonValues.get(Authority_.ROLES));
    return result;
  }

  @Override
  protected Authority buildEntityWithoutViolationEntity() {
    return buildAuthority();
  }
}

File: app/src/test/kotlin/org/artembogomolova/demo/webapp/test/db/dao/auth/AuthorityDaoTest.kt
package org.artembogomolova.demo.webapp.test.db.dao.auth

import org.artembogomolova.demo.webapp.main.dao.repo.IAuthorityRepository
import org.artembogomolova.demo.webapp.main.domain.auth.Authority
import org.artembogomolova.demo.webapp.main.domain.auth.Authority_
import org.artembogomolova.demo.webapp.main.domain.auth.Role
import org.artembogomolova.demo.webapp.main.domain.auth.User
import org.artembogomolova.demo.webapp.main.validation.UniqueMultiColumn
import org.artembogomolova.demo.webapp.test.config.AuthorityDaoTestDatasourceConfiguration
import org.artembogomolova.demo.webapp.test.db.dao.AbstractDaoTest
import org.junit.jupiter.api.DisplayName
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Import
import org.springframework.data.repository.CrudRepository

@Import(AuthorityDaoTestDatasourceConfiguration::class)
@DisplayName("Dao test: Authority")
internal class AuthorityDaoTest : AbstractDaoTest<Authority>(
    Authority::class.java,
    ::Authority
) {
    companion object {
        private const val NAME_VALUE = "test:authority"
        private const val DESCRIPTION_VALUE = "test description"
        private const val DESCRIPTION_ANOTHER_VALUE = "another description"


    }

    @Autowired
    private lateinit var authorityRepository: IAuthorityRepository
    override fun updateEntities(savedCollection: List<Authority>): List<Authority> {
        savedCollection[0].description = DESCRIPTION_ANOTHER_VALUE
        return savedCollection
    }

    override fun crudRepository(): CrudRepository<Authority, Long> = authorityRepository

    override fun generateEntities(): List<Authority> {
        return mutableListOf(buildAuthority())
    }

    private fun buildAuthority(): Authority {
        val result = Authority()
        result.name = NAME_VALUE
        result.description = DESCRIPTION_VALUE
        return result
    }

    override fun buildCommonFieldValues(uniqueMultiColumnConstraint: UniqueMultiColumn.UniqueMultiColumnConstraint): Map<String, Any> {
        val result: MutableMap<String, Any> = HashMap()
        result[Authority_.ROLES] = mutableListOf<Role>()
        result[Authority_.USERS] = mutableListOf<User>()
        return result
    }

    override fun doPrepareDeniedTestEntity(
        uniqueMultiColumnConstraint: UniqueMultiColumn.UniqueMultiColumnConstraint,
        commonValues: Map<String, Any?>
    ): Authority {
        return buildAuthority(commonValues)
    }

    override fun doDuplicateDeniedTestEntity(columns: UniqueMultiColumn.UniqueMultiColumnConstraint, commonValues: Map<String, Any?>): Authority {
        return buildAuthority(commonValues)
    }

    private fun buildAuthority(commonValues: Map<String, Any?>): Authority {
        val result = buildAuthority()
        @Suppress("UNCHECKED_CAST")
        result.roles = commonValues[Authority_.ROLES] as MutableList<Role>
        @Suppress("UNCHECKED_CAST")
        result.users = commonValues[Authority_.USERS] as MutableList<User>
        return result
    }

    override fun buildEntityWithoutViolationEntity(): Authority {
        return buildAuthority()
    }

}